zerver/views/auth.py

```diff
@@ -342,23 +346,40 @@ def oauth_redirect_to_root(request: HttpRequest, url: str,
     if next:
         params['next'] = next
 
+    params = {**params, **extra_url_params}
+
     return redirect(main_site_uri + '?' + urllib.parse.urlencode(params))
 
 def start_social_login(request: HttpRequest, backend: str) -> HttpResponse:
     backend_url = reverse('social:begin', args=[backend])
+    extra_url_params = {}  # type: Dict[str, str]
+    if backend == "saml":
+        # This backend requires the name of the IdP (from the list of configured ones)
+        # to be passed as the parameter.
+        # Currently we support configuring only one IdP.
+        # TODO: Support multiple IdPs. python-social-auth SAML (which we use here)
+        # already supports that, so essentially only the UI for it on the login pages
+        # needs to be figured out.
```

Comment:
Cool; for supporting multiple IDPs, I think it should be as simple as adjusting the URL configuration to allow passing a second argument to this method, and then tweaking the code for rendering social auth buttons to generate the right URLs as well.